src/app/codegen-data-model/CodegenDataModel_Write.cpp

```diff
@@ -0,0 +1,341 @@
+#include <lib/core/CHIPError.h>
+#include <lib/support/CodeUtils.h>
+
+#include <zap-generated/endpoint_config.h>
+
+namespace chip {
+namespace app {
+namespace {
+
+using namespace chip::app::Compatibility::Internal;
+
+/// Attempts to write via an attribute access interface (AAI)
+///
+/// If it returns a CHIP_ERROR, then this is a FINAL result (i.e. either failure or success)
+///
+/// If it returns std::nullopt, then there is no AAI to handle the given path
+/// and processing should figure out the value otherwise (generally from other ember data)
+std::optional<CHIP_ERROR> TryWriteViaAccessInterface(const ConcreteAttributePath & path, AttributeAccessInterface * aai,
+                                                     AttributeValueDecoder & decoder)
+{
+
+    // Processing can happen only if an attribute access interface actually exists..
+    if (aai == nullptr)
+    {
+        return std::nullopt;
+    }
+
+    CHIP_ERROR err = aai->Write(path, decoder);
+
+    if (err != CHIP_NO_ERROR)
+    {
+        return std::make_optional(err);
+    }
+
+    // If the decoder tried to decode, then a value should have been read for processing.
+    //   - if decode, assueme DONE (i.e. FINAL CHIP_NO_ERROR)
+    //   - if no encode, say that processing must continue
+    return decoder.TriedDecode() ? std::make_optional(CHIP_NO_ERROR) : std::nullopt;
+}
+
+/// Metadata of what a ember/pascal short string means (prepended by a u8 length)
+struct ShortPascalString
+{
+    using LengthType                        = uint8_t;
+    static constexpr LengthType kNullLength = 0xFF;
+};
+
+/// Metadata of what a ember/pascal LONG string means (prepended by a u16 length)
+struct LongPascalString
+{
+    using LengthType                        = uint16_t;
+    static constexpr LengthType kNullLength = 0xFFFF;
+};
+
+// ember assumptions ... should just work
+static_assert(sizeof(ShortPascalString::LengthType) == 1);
+static_assert(sizeof(LongPascalString::LengthType) == 2);
+
+/// Encode the value stored in 'decoder' into an ember format string 'out'
+///
+/// The value encoded will be of type T (e.g. CharSpan or ByteSpan) and it will be encoded
+/// via the given ENCODING (i.e. ShortPascalString or LongPascalString)
+///
+/// isNullable defines if the value of NULL is allowed to be encoded.
+template <typename T, class ENCODING>
+CHIP_ERROR DecodeStringLikeIntoEmberBuffer(AttributeValueDecoder decoder, bool isNullable, MutableByteSpan out)
+{
+    T workingValue;
+
+    if (isNullable)
+    {
+        typename DataModel::Nullable<T> nullableWorkingValue;
+        ReturnErrorOnFailure(decoder.Decode(nullableWorkingValue));
+
+        if (nullableWorkingValue.IsNull())
+        {
+            VerifyOrReturnError(out.size() >= sizeof(typename ENCODING::LengthType), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+            typename ENCODING::LengthType nullLength = ENCODING::kNullLength;
+            memcpy(out.data(), &nullLength, sizeof(nullLength));
```

Comment:
State that the encoding is using native byte order.

Reply:
Ember compatibility functions do a memcpy:

https://github.com/project-chip/connectedhomeip/blob/master/src/app/util/ember-compatibility-functions.cpp#L511

And then ember-strings.cpp uses LittleEndian: https://github.com/project-chip/connectedhomeip/blob/master/src/app/util/ember-strings.cpp#L34

We probably never noticed because we are not really compiling for big endian :(

I am unsure what to do here. Assuming LittleEndian is ok for now because that is what the ember code is likely using at some point. It also seems that ember-compatibility-functions is broken then.

Reply:
Will use LittleEndian since that seems explicit. But this does mean my code is different from ember-compatibility-functions (but probably more correct).